### PR TITLE
Add testcases, mock server and documentation for testing

### PR DESCRIPTION
**Pull request by SanduniU**
## Purpose
Include comprehensive testcases to test the connector with both live and mock environments.

## Changes

- Configure Mock and Live Server Setup: Set up environment configurations for live and mock servers.
- Mock Server Implementation: Added a mock server with endpoints for create, get and delete assistants, threads, messages and create and retrieve a run.
- Test Cases: Implemented test functions to validate the behavior of each endpoint, ensuring the correct response type is returned
- Include the updated dependencies 
- Add a comprehensive documentation on running tests

### CODE REVIEW

File: ballerina/tests/test.bal [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+
+configurable boolean isLiveServer = ?;
+configurable string token = ?;
+configurable string serviceUrl = isLiveServer ? "https://api.openai.com/v1" : "http://localhost:9090";
+
+ConnectionConfig config = {
+    auth: {
+        token
+    }
+};
```
Comment (ayeshLK) [inline ballerina/tests/test.bal:29]: Let's move the client initialization next to the client-config definition.
```suggestion
ConnectionConfig config = {
    auth: {
        token
    }
};
final Client openaiAssistant = check new(config,serviceUrl);
```

File: ballerina/tests/test.bal [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+
+const map<string|string[]> headers = {
+    "OpenAI-Beta": ["assistants=v2"]
+};
+
+function dataGen() returns DataPassingObject[][] {
+    return [[dataPasser]];
+}
+
+@test:Config {dataProvider:  dataGen, dependsOn: [testCreateMessage, testCreateThread, testGetMessage]}
```
Comment (ayeshLK) [inline ballerina/tests/test.bal]: Shall we format this code properly ? You can do this using the Ballerina CLI command.
- Go into the `ballerina` directory.
- Execute `bal format tests/tests.bal`

File: ballerina/tests/test.bal [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+isolated function  testDeleteMessage(DataPassingObject dataPasser) {
+    // Ensure threadId and messageId are not empty
```
Comment (ayeshLK) [inline ballerina/tests/test.bal]: Lets remove the comments from the test cases as they are redundant. Check for other places as well.

File: ballerina/tests/test.bal [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+    }
+
+    // Delete the message
+    DeleteMessageResponse|error res = AssistantClient->/threads/[dataPasser.threadId]/messages/[dataPasser.messageId].delete(headers);
+    if (res is DeleteMessageResponse) {
+        io:println("Message deleted successfully: ", res);
+        test:assertTrue(res.deleted == true, msg = "Failed to delete message");
+    } else {
+        test:assertFail(msg = "Failed to delete message");
+    }
```
Comment (ayeshLK) [inline ballerina/tests/test.bal]: Lets follow this style in the other places as well.
```suggestion
    DeleteMessageResponse res = check AssistantClient->/threads/[dataPasser.threadId]/messages/[dataPasser.messageId].delete(headers);
    io:println("Message deleted successfully: ", res);
    test:assertTrue(res.deleted == true, msg = "Failed to delete message");
```

Comment (SanduniU) [reply ballerina/tests/test.bal]: The resource functions are declared with union types in the client.bal. Is there any way to check the fields of the response?

Comment (ayeshLK) [reply ballerina/tests/test.bal]: I think this approach fix the problem. Shall we do this and see ?

File: ballerina/tests/mock_server.bal [hunk 5]
```diff
@@ -0,0 +1,993 @@
...
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import ballerina/http;
+import ballerina/log;
+
+listener http:Listener httpListener = new (9090);
+
```
Comment (ayeshLK) [inline ballerina/tests/mock_server.bal]:
```suggestion
```

**Comment by ayeshLK** [comment]
Lets rename `mock_server.bal` as `mock_service.bal`.

File: ballerina/tests/test.bal (hunk 1 above)
Comment (ayeshLK) [reply ballerina/tests/test.bal:29]: Shall we rename `AssistantClient` to `openaiAssistant` ?

File: ballerina/tests/test.bal (hunk 2 above)
Comment (ayeshLK) [reply ballerina/tests/test.bal]: Can you reformat the test code ?

File: ballerina/tests/test.bal [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+
+const map<string|string[]> headers = {
+    "OpenAI-Beta": ["assistants=v2"]
+};
+
+function dataGen() returns DataPassingObject[][] {
+    return [[dataPasser]];
+}
+
+@test:Config {dataProvider: dataGen, dependsOn: [testCreateMessage, testCreateThread, testGetMessage]}
```
Comment (NipunaRanasinghe) [inline ballerina/tests/test.bal]: Any particular reason to use a data provider for this test suite (as test data in L40 seems to be empty)? 

If not, better to completely remove the data provider related implementations

File: ballerina/tests/test.bal [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+        test:assertFail(msg = "No thread ID or message ID available. Ensure message creation test runs first.");
+    }
+
+    DeleteMessageResponse res = check openAIAssistant->/threads/[dataPasser.threadId]/messages/[dataPasser.messageId].delete(headers);
+    io:println("Message deleted successfully: ", res);
+    test:assertTrue(res.deleted == true, msg = "Failed to delete message");
+
+}
+
+@test:Config {dataProvider: dataGen, dependsOn: [testCreateThread, testCreateMessage]}
```
Comment (NipunaRanasinghe) [inline ballerina/tests/test.bal]: This is the usual formatting style used for annotations
```suggestion
@test:Config {
    dataProvider: dataGen, 
    dependsOn: [testCreateThread, testCreateMessage]
}
```

File: ballerina/tests/test.bal [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+};
+
+function dataGen() returns DataPassingObject[][] {
+    return [[dataPasser]];
+}
+
+@test:Config {dataProvider: dataGen, dependsOn: [testCreateMessage, testCreateThread, testGetMessage]}
+isolated function testDeleteMessage(DataPassingObject dataPasser) returns error? {
+
+    if (dataPasser.threadId == "" || dataPasser.messageId == "") {
```
Comment (NipunaRanasinghe) [inline ballerina/tests/test.bal]: No need to use parenthesis for if statements in Ballerina (https://learn-ballerina.github.io/best_practices/avoid_parentheses.html)
```suggestion
    if dataPasser.threadId == "" || dataPasser.messageId == "" {
```

File: ballerina/tests/test.bal [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+    ListAssistantsQueries query = {
+        before: "",
+        after: ""
+    };
+
+    ListAssistantsResponse res = check openAIAssistant->/assistants.get(headers, query);
+    io:println("Assistant List: ", res.data);
+    test:assertTrue(res is ListAssistantsResponse);
+
+}
```
Comment (NipunaRanasinghe) [inline ballerina/tests/test.bal]: additional newlines. please fix all the similar places
```suggestion
isolated function testListAssistants() returns error? {
    ListAssistantsQueries query = {
        before: "",
        after: ""
    };

    ListAssistantsResponse res = check openAIAssistant->/assistants.get(headers, query);
    io:println("Assistant List: ", res.data);
    test:assertTrue(res is ListAssistantsResponse);
}
```

File: ballerina/tests/test.bal [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+        test:assertFail(msg = "No thread ID or message ID available. Ensure thread creation and message creation tests run first.");
+    }
+
+    MessageObject res = check openAIAssistant->/threads/[dataPasser.threadId]/messages/[dataPasser.messageId].get(headers);
+    io:println("Message Details: ", res);
+    test:assertEquals(res.id, dataPasser.messageId, msg = "Retrieved message ID does not match the requested ID");
+
+}
+
+@test:Config {dataProvider: dataGen, dependsOn: [testCreateThread]}
```
Comment (NipunaRanasinghe) [inline ballerina/tests/test.bal]: missing test groups  `mock_tests`  and `live_tests`. Please update all the test cases with the same

File: ballerina/tests/test.bal [hunk 11]
```diff
@@ -0,0 +1,262 @@
...
+string threadId = "";
+string messageId = "";
+string runId = "";
+string stepId = "";
+
+const map<string|string[]> headers = {
+    "OpenAI-Beta": ["assistants=v2"]
+};
+
+@test:Config {dependsOn: [testCreateMessage, testCreateThread, testGetMessage]}
```
Comment (NipunaRanasinghe) [inline ballerina/tests/test.bal]: Usually we keep the annnotations in this format. Shall we stick to that?
```suggestion
@test:Config {
    dependsOn: [testCreateMessage, testCreateThread, testGetMessage]
}
```

File: ballerina/tests/test.bal [hunk 12]
```diff
@@ -0,0 +1,262 @@
...
+    //     max_completion_tokens: 200
+    // };
+
+    // RunObject resp = check openAIAssistant->/threads/runs.post(createThreadAndRunReq, headers);
+    // io:println("Created Thread and Run: ", resp);
+    // test:assertNotEquals(resp.id, "", msg = "Thread and Run creation failed: No Run ID returned");
+
+}
+
+@test:Config {}
```
Comment (NipunaRanasinghe) [inline ballerina/tests/test.bal]: Please updates all the test cases with the proper test groups (`live_tests` and `mock_tests`)

**Review by NipunaRanasinghe — APPROVED** [review]
LGTM

**Pull request merged**